docs/remctld.8.in
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "REMCTLD 8"
.TH REMCTLD 8 "2022-05-09" "3.18" "remctl"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
remctld \- Server for remctl, a remote command execution utility
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
remctld [\fB\-dFhmSvZ\fR] [\fB\-b\fR \fIbind-address\fR [\fB\-b\fR \fIbind-address\fR ...]]
    [\fB\-f\fR \fIconfig\fR] [\fB\-k\fR \fIkeytab\fR] [\fB\-P\fR \fIfile\fR] [\fB\-p\fR \fIport\fR]
    [\fB\-s\fR \fIservice\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBremctld\fR is the server for remctl.  It accepts a connection from remctl,
receives the command to execute and the arguments, verifies authorization
of the user and executes the command, returning the result back to the
client.  All connections are authenticated using Kerberos GSS-API
Kerberos, and all transmissions are also encrypted using the GSS-API
privacy layer.
.PP
\&\fBremctld\fR is normally started using \fBtcpserver\fR or from \fBinetd\fR, but it
may be run in stand-alone mode as a daemon using \fB\-m\fR.  Either \fB\-s\fR must
be given to use an alternate identity (which will require the same flag be
used for \fBremctl\fR client invocations), or it must be run as root to read
the host keytab file.  \fBremctld\fR logs its activity using syslog (the
daemon facility).
.PP
The location of the configuration file may be specified with the \fB\-f\fR
option.  The default location is \fI\f(CI@sysconfdir\fI@/remctl.conf\fR.  For
information on the format of the configuration file, see \*(L"\s-1CONFIGURATION
FILE\*(R"\s0 below.
.PP
When the command is run, several environment variables will be set
providing information about the remote connection.  See \s-1ENVIRONMENT\s0
below for more information.
.PP
Command-line and configuration options and \s-1ACL\s0 methods are annotated below
with the version at which they were added.  For version information for
more general features, see \s-1COMPATIBILITY\s0 below.
.SH "OPTIONS"
.IX Header "OPTIONS"
The start of each option description is annotated with the version of
\&\fBremctld\fR in which that option was added with its current meaning.
.IP "\fB\-b\fR \fIbind-address\fR" 4
.IX Item "-b bind-address"
[2.17] When running as a standalone server, bind to the specified local
address rather than listening on all interfaces.  This option may be given
multiple times to bind to multiple addresses.  \fIbind-address\fR must be an
\&\s-1IP\s0 address (either IPv4 or IPv6), not a hostname.  Only makes sense in
combination with \fB\-m\fR.
.Sp
This option is ignored if \fBremctld\fR is passed already open sockets via
the systemd socket activation protocol.  In that case, the bind addresses
of the sockets should be controlled via the systemd configuration.
.IP "\fB\-d\fR" 4
.IX Item "-d"
[1.10] Enable verbose debug logging to syslog (or to standard output if
\&\fB\-S\fR is also given).
.IP "\fB\-F\fR" 4
.IX Item "-F"
[2.8] Normally when running in stand-alone mode (\fB\-m\fR), \fBremctld\fR
backgrounds itself to run as a daemon, changes directory to \fI/\fR, and
drops any controlling terminal.  This flag suppresses this behavior,
usually for debugging or so that \fBremctld\fR can be monitored by other
processes.
.IP "\fB\-f\fR \fIconfig\fR" 4
.IX Item "-f config"
[1.0] The configuration file for \fBremctld\fR, overriding the default path.
.IP "\fB\-h\fR" 4
.IX Item "-h"
[1.10] Show a brief usage message and then exit.  This usage method will
include a list of supported \s-1ACL\s0 types and can be used to determine if
optional \s-1ACL\s0 methods were compiled into a given \fBremctld\fR build.
.IP "\fB\-k\fR \fIkeytab\fR" 4
.IX Item "-k keytab"
[2.8] Use \fIkeytab\fR as the keytab for server credentials rather than the
system default or the value of the \s-1KRB5_KTNAME\s0 environment variable.
Using \fB\-k\fR just sets the \s-1KRB5_KTNAME\s0 environment variable internally in
the process.
.IP "\fB\-m\fR" 4
.IX Item "-m"
[2.8] Enable stand-alone mode.  \fBremctld\fR will listen to its configured
port and fork a new child for each incoming connection.  By default, when
this option is used, \fBremctld\fR also changes directory to \fI/\fR,
backgrounds itself, and closes standard input, output, and error.  To not
background, pass \fB\-F\fR as well.  To not close standard output and error
and continue using them for logging, pass \fB\-S\fR as well.
.Sp
To determine the port, \fBremctld\fR attempts to look up the \f(CW\*(C`remctl\*(C'\fR
service in the local \fI/etc/services\fR file and uses the port defined
there.  If the \f(CW\*(C`remctl\*(C'\fR service could not be found, it uses 4373, the
registered remctl port.
.Sp
When running in stand-alone mode, send the \s-1SIGHUP\s0 signal to \fBremctld\fR to
ask it to re-read its configuration file and \s-1SIGTERM\s0 to ask it to exit.
.IP "\fB\-P\fR \fIfile\fR" 4
.IX Item "-P file"
[2.0] When running in stand-alone mode (\fB\-m\fR), write the \s-1PID\s0 of
\&\fBremctld\fR to \fIfile\fR.  This option is ignored unless \fB\-m\fR is also given.
.IP "\fB\-p\fR \fIport\fR" 4
.IX Item "-p port"
[1.0] When running in stand-alone mode, listen on port \fIport\fR rather than
the default.  This option does nothing unless used with \fB\-m\fR.
.Sp
This option is ignored if \fBremctld\fR is passed already open sockets via

NEWS
                       User-Visible remctl Changes

remctl 3.18 (2022-05-08)

    Add support for PCRE2 for pcre ACLs and use it by preference over
    PCRE1 if it is found.  UTF-8 regular expressions are not enabled by
    default, but can be enabled by adding (*UTF) to the beginning of the
    regular expression (a standard PCRE2 feature not specific to remctl).

    Document that pcre and regex ACL expressions are not automatically
    anchored at the start and end of the principal name, so they should be
    explicitly anchored in the configuration unless one intends to allow
    partial matches.

    The Perl libraries and the remctl test suite now require Perl 5.10 or
    later.

    Document sending SIGHUP to remctld when running in stand-alone mode to
    ask it to re-read its configuration file, and document that SIGTERM
    will cause it to exit.  (Fixes #30)

    Mark remctl client library functions that allocate memory with their
    corresponding deallocation functions so that GCC 11 and later can
    diagnose memory deallocation bugs.

    Remove remaining references to pytest-runner in the Python bindings.
    Thanks, Ken Dreyer.

    Switch the Ruby bindings tests to Minitest from Test::Unit.  Thanks,
    Ken Dreyer.

    Update to rra-c-util 10.2:

    * Fix IN6_ARE_ADDR_EQUAL Autoconf probe on macOS.
    * Fix compiler flag probes with Clang.
    * Check that Kerberos headers can be included.
    * Fix util/fdflag test on Solaris 11 and OpenBSD.
    * Suppress some additional warnings in GCC 10.

remctl 3.17 (2020-12-13)

    Port the PHP extention to PHP 8.  This required declaring the
    arguments to the functions (which should have been done with PHP 7)
    and removing some obsolete constructs.

    Make the Python install_requires dependency on typing conditional on
    Python versions earlier than 3.5 so that setuptools won't attempt to
    download typing when it's part of the standard library.  Thanks to
    Gianfranco Costamagna and Matthias Klose for the bug report.

    Fix the Python module build to more reliably test the newly-built
    module and to enable verbose testing.

    Fix non-Kerberos network tests on hosts with no IPv4 addresses.  In
    this case, the network tests for binding all configured addresses will
    bind only to IPv6, which broke some prior assumptions in the test
    suite.  Thanks to Niko Tyni for the bug report.  Note that the tests
    that require a Kerberos setup will still fail in this scenario, since
    they assume remctld will bind to 127.0.0.1 by default.

    Stop providing a replacement for a broken snprintf and assume the libc
    version works correctly.  This portability code has proven difficult
    to maintain, and was only relevant for ancient proprietary UNIX
    versions that have been obsolete for many years.

    Update to rra-c-util 8.4:

    * Fix reallocarray prototyping on NetBSD.
    * Fix getnameinfo tests on musl-based Linux distributions.
    * Include string.h when probing for getaddrinfo properties.
    * Fix Perl style issues found by Perl::Critic::Freenode.
    * Fix support for configuring the test suite with a krb5.conf file.
    * Fix tests when the system krb5.conf file does not set default_realm.
    * Ignore files in tests/config when checking for license identifiers.
    * Ignore object files when checking for license identifiers.
    * Drop support for Perl 5.6.
    * Reformat all C source using clang-format 10.
    * Remove bogus snprintf tests.

    Update to C TAP Harness 4.7:

    * Fix warning with GCC 10.

remctl 3.16 (2019-10-26)

    The Python bindings now support Python 3.  They have been tested only
    with Python 2.7 and Python 3.7, but should work with any version of
    Python 3 later than Python 3.1.

    The Python bindings have been modernized to remove obsolete syntax,
    which may mean that versions of Python back to Python 2.3 are no
    longer supported.  The bindings are only tested with Python 2.7.
    Passing in anything other than an iterable of str or bytes as the
    command to run is deprecated and support for using objects that can be
    converted to str in commands will be removed in a future release.

    Update the Python bindings documentation to use proper Python types
    and to document how str and bytes values are handled.

    Building the Python bindings now requires the setuptools, pytest,
    pytest-runner, and typing Python modules.

    Add -t flag to the remctl client to specify the network timeout.
    Thanks, Remi Ferrand.


server/remctld.c
/*
 * The remctl server.
 *
 * Handles option parsing, network setup, and the basic processing loop of the
 * remctld server.  Supports either being run from inetd or tcpserver or
 * running as a stand-alone daemon and managing its own network connections.
 *
 * Written by Anton Ushakov
 * Extensive modifications by Russ Allbery <eagle@eyrie.org>
 * Copyright 2018, 2020 Russ Allbery <eagle@eyrie.org>
 * Copyright 2002-2008, 2010-2012, 2014
 *     The Board of Trustees of the Leland Stanford Junior University
 *
 * SPDX-License-Identifier: MIT
 */

#include <config.h>
#include <portable/event.h>
#include <portable/gssapi.h>
#include <portable/sd-daemon.h>
#include <portable/socket.h>
#include <portable/system.h>

#include <signal.h>
#include <sys/wait.h>
#include <syslog.h>
#include <time.h>

#include <server/internal.h>
#include <util/fdflag.h>
#include <util/macros.h>
#include <util/messages.h>
#include <util/network.h>
#include <util/vector.h>
#include <util/xmalloc.h>

/*
 * Flag indicating whether we've received a SIGCHLD and need to reap children
 * (only used in standalone mode).
 */
static volatile sig_atomic_t child_signaled = 0;

/*
 * Flag indicating whether we've received a signal asking us to re-read our
 * configuration file (only used in standalone mode).
 */
static volatile sig_atomic_t config_signaled = 0;

/*
 * Flag indicating whether we've received a signal asking us to exit (only
 * used in standalone mode).
 */
static volatile sig_atomic_t exit_signaled = 0;

/* Usage message. */
static const char usage_message[] = "\
Usage: remctld <options>\n\
\n\
Options:\n\
    -b <addr>     Bind to a specific address (may be given multiple times)\n\
    -d            Log verbose debugging information\n\
    -F            Run in the foreground instead of forking and exiting\n\
    -f <file>     Config file (default: " CONFIG_FILE ")\n\
    -h            Display this help\n\
    -m            Stand-alone daemon mode, meant mostly for testing\n\
    -P <file>     Write PID to file, only useful with -m\n\
    -p <port>     Port to use, only for standalone mode (default: 4373)\n\
    -S            Log to standard output/error rather than syslog\n\
    -s <service>  Service principal to use (default: host/<host>)\n\
    -v            Display the version of remctld\n\
    -Z            Raise SIGSTOP once ready for connections\n\
\n\
Supported ACL methods: file, princ, deny";

/* Structure used to store program options. */
struct options {
    bool debug;               /* -d: log verbose debugging information */
    bool foreground;          /* -F: run in the foreground */
    bool log_stdout;          /* -S: log to standard output and error */
    bool standalone;          /* -m: run in stand-alone daemon mode */
    bool suspend;             /* -Z: raise SIGSTOP when ready */
    unsigned short port;      /* -p: port on which to listen */
    char *service;            /* -s: service principal to use */
    const char *config_path;  /* -f: path to the configuration file */
    const char *pid_path;     /* -P: path to the PID file to write */
    struct vector *bindaddrs; /* -b: bind to a specific address */
};


/*
 * Display the usage message for remctld.
 */
__attribute__((__noreturn__)) static void
usage(int status)
{
    FILE *output;

    output = (status == 0) ? stdout : stderr;
    if (status != 0)
        fprintf(output, "\n");
    fprintf(output, usage_message);
#ifdef HAVE_GPUT
    fprintf(output, ", gput");
#endif
#if defined(HAVE_KRB5) && defined(HAVE_GETGRNAM_R)
    fprintf(output, ", localgroup");
#endif
#ifdef HAVE_PCRE
    fprintf(output, ", pcre");
#endif
#ifdef HAVE_REGCOMP
    fprintf(output, ", regex");
#endif
    fprintf(output, "\n");
    exit(status);
}


/*
 * Signal handler for child processes forked when running in standalone mode.
 * Just set the child_signaled global so that we know to reap the processes
 * later.
 */
static void
child_handler(int sig UNUSED)
{
    child_signaled = 1;
}


/*
 * Signal handler for signals asking us to re-read our configuration file when
 * running in standalone mode.  Set the config_signaled global so that we do
 * this the next time through the processing loop.
 */
static void
config_handler(int sig UNUSED)
{
    config_signaled = 1;
}


/*
 * Signal handler for signals asking for a clean shutdown when running in
 * standalone mode.  Set the exit_signaled global so that we exit cleanly the
 * next time through the processing loop.
 */
static void
exit_handler(int sig UNUSED)
{
    exit_signaled = 1;
}


/*
 * Given a service name, imports it and acquires credentials for it, storing
 * them in the second argument.  Returns true on success and false on failure,
 * logging an error message.
 *
 * Normally, you don't want to do this; instead, normally you want to allow
 * the underlying GSS-API library choose the appropriate credentials from a
 * keytab for each incoming connection.
 */
static bool
acquire_creds(char *service, gss_cred_id_t *creds)
{
    gss_buffer_desc buffer;
    gss_name_t name;
    OM_uint32 major, minor;

    buffer.value = service;
    buffer.length = strlen(buffer.value) + 1;
    major = gss_import_name(&minor, &buffer, GSS_C_NT_USER_NAME, &name);
    if (major != GSS_S_COMPLETE) {
        warn_gssapi("while importing name", major, minor);
        return false;
    }
    major = gss_acquire_cred(&minor, name, 0, GSS_C_NULL_OID_SET, GSS_C_ACCEPT,
                             creds, NULL, NULL);
    if (major != GSS_S_COMPLETE) {
        warn_gssapi("while acquiring credentials", major, minor);
        return false;
    }
    gss_release_name(&minor, &name);
    return true;
}


/*
 * Handle the interaction with the client.  Takes the client file descriptor,
 * the server configuration, and the server credentials.  Establishes a
 * security context, processes requests from the client, checks the ACL file
 * as appropriate, and then spawns commands, sending the output back to the
 * client.  This function only returns when the client connection has
 * completed, either successfully or unsuccessfully.
 */
static void
handle_connection(int fd, struct config *config, gss_cred_id_t creds)
{
    struct client *client;

    /* Establish a context with the client. */
    client = server_new_client(fd, creds);
    if (client == NULL) {
        close(fd);
        return;
    }
    debug("accepted connection from %s (protocol %d)", client->user,
          client->protocol);

    /*
     * Now, we process incoming commands.  This is handled differently
     * depending on the protocol version.  These functions won't exit until
     * the client is done sending commands and we're done replying.
     */
    if (client->protocol == 1)
        server_v1_handle_messages(client, config);
    else
        server_v2_handle_messages(client, config);

    /* We're done; shut down the client connection. */
    server_free_client(client);
}


/*
 * Gather information about an exited child and log an appropriate message.
 * We keep the log level to debug unless something interesting happened, like
 * a non-zero exit status.
 */
static void
log_child(pid_t pid, int status)
{
    if (WIFEXITED(status)) {
        if (WEXITSTATUS(status) != 0)
            warn("child %lu exited with %d", (unsigned long) pid,
                 WEXITSTATUS(status));
        else
            debug("child %lu done", (unsigned long) pid);
    } else if (WIFSIGNALED(status)) {
        warn("child %lu died on signal %d", (unsigned long) pid,
             WTERMSIG(status));
    } else {
        warn("child %lu died", (unsigned long) pid);
    }
}


/*
 * Given a bind address, return true if it's an IPv6 address.  Otherwise, it's
 * assumed to be an IPv4 address.
 */
#ifdef HAVE_INET6
static bool
is_ipv6(const char *string)
{
    struct in6_addr addr;
    return inet_pton(AF_INET6, string, &addr) == 1;
}
#else
static bool
is_ipv6(const char *string UNUSED)
{
    return false;
}
#endif


/*
 * Bind the listening socket or sockets on which we accept requests and return
 * a list of sockets in the fds parameter.  Return a count of sockets in the
 * count parameter.
 *
 * Handle the socket activation case where the socket has already been set up
 * for us by systemd and, in that case, just return the already-configured
 * socket.
 */
static void
bind_sockets(struct options *options, socket_type **fds, unsigned int *count)
{
    int status, fd_index;
    size_t i;
    const char *addr;
    socket_type fd;

    /* Check whether systemd has already bound the sockets. */
    status = sd_listen_fds(true);
    if (status < 0)
        die("using systemd-bound sockets failed: %s", strerror(-status));
    if (status > 0) {
        *fds = xcalloc(status, sizeof(socket_type));
        for (fd_index = 0; fd_index < status; fd_index++)
            (*fds)[fd_index] = SD_LISTEN_FDS_START + fd_index;
        *count = status;
        return;
    }

    /*
     * We have to do the work ourselves.  If there is no bind address, bind to
     * all local sockets, which will normally result in two file descriptors
     * on which to listen.
     */
    if (options->bindaddrs->count == 0) {
        if (!network_bind_all(SOCK_STREAM, options->port, fds, count))
            sysdie("cannot bind any sockets");
        for (fd_index = 0; fd_index < (int) *count; fd_index++)
            if (listen((*fds)[fd_index], 5) < 0)
                sysdie("error listening on socket");
        return;
    }

    /*
     * Otherwise, we have to iterate through all the bind addresses, bind them
     * using the appropriate function, and listen on each.
     */
    *count = (unsigned int) options->bindaddrs->count;
    *fds = xcalloc(*count, sizeof(socket_type));
    for (i = 0; i < options->bindaddrs->count; i++) {
        addr = options->bindaddrs->strings[i];
        if (is_ipv6(addr))
            fd = network_bind_ipv6(SOCK_STREAM, addr, options->port);
        else
            fd = network_bind_ipv4(SOCK_STREAM, addr, options->port);
        if (fd == INVALID_SOCKET)
            sysdie("cannot bind to address %s, port %hu", addr, options->port);
        if (listen(fd, 5) < 0)
            sysdie("error listening on socket");
        (*fds)[i] = fd;
    }
}


/*
 * Write a PID to a file.  This is done via atomic replacement so that the
 * file never exists with no content.  Note that there is no locking and no
 * verification that an existing PID has exited.
 */
static void
write_pidfile(pid_t pid, const char *path)
{
    char *template;
    FILE *pid_file;
    int fd;

    xasprintf(&template, "%s.XXXXXX", path);
    fd = mkstemp(template);
    if (fd < 0)
        sysdie("cannot create temporary PID file %s", template);
    pid_file = fdopen(fd, "w");
    if (pid_file == NULL)
        sysdie("cannot reopen temporary PID file %s", template);
    if (fprintf(pid_file, "%ld\n", (long) pid) < 0)
        sysdie("cannot write to temporary PID file %s", template);
    fclose(pid_file);
    if (rename(template, path) < 0)
        sysdie("cannot rename temporary PID file to %s", path);
    free(template);
}


/*
 * Run as a daemon.  This is the main dispatch loop, which listens for network
 * connections, forks a child to process each connection, and reaps the
 * children when they're done.  This is only used in standalone mode; when run
 * from inetd or tcpserver, remctld processes one connection and then exits.
 */
static void
server_daemon(struct options *options, struct config *config,
              gss_cred_id_t creds)
{
    socket_type s;
    unsigned int nfds, i;
    socket_type *fds;
    pid_t child;
    int status;
    struct sigaction sa, oldsa;
    struct sockaddr_storage ss;
    socklen_t sslen;
    char ip[INET6_ADDRSTRLEN];
    OM_uint32 minor;

    /* Set up a SIGCHLD handler so that we know when to reap children. */
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = child_handler;
    if (sigaction(SIGCHLD, &sa, &oldsa) < 0)
        sysdie("cannot set SIGCHLD handler");

    /* Set up exit handlers for signals that call for a clean shutdown. */
    sa.sa_handler = exit_handler;
    if (sigaction(SIGINT, &sa, NULL) < 0)
        sysdie("cannot set SIGINT handler");
    if (sigaction(SIGTERM, &sa, NULL) < 0)
        sysdie("cannot set SIGTERM handler");

    /* Set up a SIGHUP handler so that we know when to re-read our config. */
    sa.sa_handler = config_handler;
    if (sigaction(SIGHUP, &sa, NULL) < 0)
        sysdie("cannot set SIGHUP handler");

    /* Bind to the network sockets and configure listening addresses. */
    bind_sockets(options, &fds, &nfds);

    /*
     * Set up our PID file now that we're ready to accept connections, so that
     * the PID file isn't created until clients can connect.
     */
    if (options->pid_path != NULL)
        write_pidfile(getpid(), options->pid_path);

    /* Log a starting message. */
    notice("starting");

    /* Indicate to systemd that we're ready to answer requests. */
    status = sd_notify(true, "READY=1");
    if (status < 0)
        warn("cannot notify systemd of startup: %s", strerror(-status));

    /* Indicate to upstart that we're ready to answer requests. */
    if (options->suspend)
        if (raise(SIGSTOP) < 0)
            syswarn("cannot notify upstart of startup");

    /*
     * The main processing loop.  Each time through the loop, check to see if
     * we need to reap children, check to see if we should re-read our
     * configuration, and check to see if we're exiting.  Then see if we have
     * a new connection, and if so, fork a child to handle it.
     *
     * Note that there are no limits here on the number of simultaneous
     * processes, so you may want to set system resource limits to prevent an
     * attacker from consuming all available processes.
     */
    while (1) {
        if (child_signaled) {
            child_signaled = 0;
            while ((child = waitpid(0, &status, WNOHANG)) > 0)
                log_child(child, status);
            if (child < 0 && errno != ECHILD)
                sysdie("waitpid failed");
        }
        if (config_signaled) {
            config_signaled = 0;
            notice("re-reading configuration");
            server_config_free(config);
            config = server_config_load(options->config_path);
            if (config == NULL)
                die("cannot load configuration file %s", options->config_path);
        }
        if (exit_signaled) {
            notice("signal received, exiting");
            break;
        }
        sslen = sizeof(ss);
        s = network_accept_any(fds, nfds, (struct sockaddr *) &ss, &sslen);
        if (s == INVALID_SOCKET) {
            if (errno != EINTR)
                sysdie("error accepting incoming connection");
            continue;
        }
        fdflag_close_exec(s, true);
        child = fork();
        if (child < 0) {
            syswarn("forking a new child failed");
            warn("sleeping ten seconds in the hope we recover...");
            sleep(10);
        } else if (child == 0) {
            for (i = 0; i < nfds; i++)
                close(fds[i]);
            network_bind_all_free(fds);
            if (sigaction(SIGCHLD, &oldsa, NULL) < 0)
                syswarn("cannot reset SIGCHLD handler");
            handle_connection(s, config, creds);
            if (creds != GSS_C_NO_CREDENTIAL)
                gss_release_cred(&minor, &creds);
            if (options->log_stdout)
                fflush(stdout);
            server_config_free(config);
            vector_free(options->bindaddrs);
            libevent_global_shutdown();
            message_handlers_reset();
            exit(0);
        } else {
            close(s);
            network_sockaddr_sprint(ip, sizeof(ip), (struct sockaddr *) &ss);
            debug("child %lu for %s", (unsigned long) child, ip);
        }
    }

    /*
     * Clean up resources at the end of the loop.  This is not strictly
     * necessary, but it helps valgrind testing.
     */
    if (options->pid_path != NULL)
        unlink(options->pid_path);
    for (i = 0; i < nfds; i++)
        close(fds[i]);
    network_bind_all_free(fds);
}


/*
 * Main routine.  Parses command-line arguments, determines whether we're
 * running in stand-alone or inetd mode, and does the connection handling if
 * running in standalone mode.  User connections are handed off to
 * process_connection.
 */
int
main(int argc, char *argv[])
{
    struct options options;
    int option;
    long tmp_port;
    char *end;
    struct sigaction sa;
    gss_cred_id_t creds = GSS_C_NO_CREDENTIAL;
    OM_uint32 minor;
    struct config *config;

    /* Ignore SIGPIPE errors from our children. */
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = SIG_IGN;
    if (sigaction(SIGPIPE, &sa, NULL) < 0)
        sysdie("cannot set SIGPIPE handler");

    /* Establish identity for logging. */
    message_program_name = "remctld";

    /* Initialize the logging and fatal callbacks for libevent. */
    event_set_log_callback(server_event_log_callback);
    event_set_fatal_callback(server_event_fatal_callback);

    /* Initialize options. */
    memset(&options, 0, sizeof(options));
    options.port = 4373;
    options.config_path = CONFIG_FILE;
    options.bindaddrs = vector_new();

    /* Parse options. */
    while ((option = getopt(argc, argv, "b:dFf:hk:mP:p:Ss:vZ")) != EOF) {
        switch (option) {
        case 'b':
            vector_add(options.bindaddrs, optarg);
            break;
        case 'd':
            options.debug = true;
            break;
        case 'F':
            options.foreground = true;
            break;
        case 'f':
            options.config_path = optarg;
            break;
        case 'h':
            usage(0);
        case 'k':
            if (setenv("KRB5_KTNAME", optarg, 1) < 0)
                sysdie("cannot set KRB5_KTNAME");
            break;
        case 'm':
            options.standalone = true;
            break;
        case 'P':
            options.pid_path = optarg;
            break;
        case 'p':
            tmp_port = strtol(optarg, &end, 10);
            if (*end != '\0' || tmp_port < 1 || tmp_port > (1L << 16) - 1)
                die("invalid port number %ld", tmp_port);
            options.port = (unsigned short) tmp_port;
            break;
        case 'S':
            options.log_stdout = true;
            break;
        case 's':
            options.service = optarg;
            break;
        case 'v':
            printf("remctld %s\n", PACKAGE_VERSION);
            exit(0);
        case 'Z':
            options.suspend = true;
            break;
        default:
            usage(1);
        }
    }

    /* Check arguments for consistency. */
    if (options.bindaddrs->count > 0 && !options.standalone)
        die("-b only makes sense in combination with -m");
    if (options.suspend && !options.standalone)
        die("-Z only makes sense in combination with -m");

    /* Daemonize if told to do so. */
    if (options.standalone && !options.foreground)
        if (daemon(0, options.log_stdout) != 0)
            sysdie("cannot daemonize");

    /*
     * Set up syslog unless stdout/stderr was requested.  Set up debug logging
     * if requested.
     */
    if (options.log_stdout) {
        if (options.debug)
            message_handlers_debug(1, message_log_stdout);
    } else {
        openlog("remctld", LOG_PID | LOG_NDELAY, LOG_DAEMON);
        message_handlers_notice(1, message_log_syslog_info);
        message_handlers_warn(1, message_log_syslog_warning);
        message_handlers_die(1, message_log_syslog_err);
        if (options.debug)
            message_handlers_debug(1, message_log_syslog_debug);
    }

    /* Read the configuration file. */
    config = server_config_load(options.config_path);
    if (config == NULL)
        die("cannot read configuration file %s", options.config_path);

    /*
     * If a service was specified, we should load only those credentials since
     * those are the only ones we're allowed to use.  Otherwise, creds will
     * keep its default value of GSS_C_NO_CREDENTIAL, which means support
     * anything that's in the keytab.
     */
    if (options.service != NULL)
        if (!acquire_creds(options.service, &creds))
            die("unable to acquire creds, aborting");

    /*
     * If we're not running as a daemon, just process the connection.
     * Otherwise, create a socket and listen on the socket, processing each
     * incoming connection.
     */
    if (!options.standalone)
        handle_connection(STDIN_FILENO, config, creds);
    else
        server_daemon(&options, config, creds);

    /* Clean up and exit. */
    server_config_free(config);
    if (creds != GSS_C_NO_CREDENTIAL)
        gss_release_cred(&minor, &creds);
    vector_free(options.bindaddrs);
    libevent_global_shutdown();
    message_handlers_reset();
    return 0;
}
